modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java

```diff
@@ -1945,21 +1947,25 @@ private void handleArrayType(ArrayMapAccessExpr arrayMapAccessExpr) {
         if (arrayMapVarRefExpr.getType() instanceof BArrayType) {
 
             // Check the type of the index expression
-            Expression indexExpr = arrayMapAccessExpr.getIndexExpr();
-            visitSingleValueExpr(indexExpr);
-            if (indexExpr.getType() != BTypes.typeInt) {
-                BLangExceptionHelper.throwSemanticError(arrayMapAccessExpr, SemanticErrors.NON_INTEGER_ARRAY_INDEX,
-                        indexExpr.getType());
+            for (Expression indexExpr : arrayMapAccessExpr.getIndexExpr()) {
+                visitSingleValueExpr(indexExpr);
+                if (indexExpr.getType() != BTypes.typeInt) {
+                    BLangExceptionHelper.throwSemanticError(arrayMapAccessExpr, SemanticErrors.NON_INTEGER_ARRAY_INDEX,
+                            indexExpr.getType());
+                }
             }
 
             // Set type of the arrays access expression
-            BType typeOfArray = ((BArrayType) arrayMapVarRefExpr.getType()).getElementType();
-            arrayMapAccessExpr.setType(typeOfArray);
+            BType expectedType =  arrayMapVarRefExpr.getType();
+            for (int i = 0; i < arrayMapAccessExpr.getIndexExpr().length; i++) {
+                expectedType =  ((BArrayType) expectedType).getElementType();
+            }
+            arrayMapAccessExpr.setType(expectedType);
 
```

Comment:
Here ArrayMapAccessExpr itself does not have a type. The type is taken from VariableRefExpr. VariableRefExpr's type is taken by looking up against the variable name. 

IMO, even if ArrayMapAccessExpr has the type, it won't be helpful. For example let's say,

```
int[][][][] xxxx = [threeDarray];
int i = xxxx[0][0][0][0];
```

In the above case, suppose type of ArrayMapAccessExpr is int[][][][]. Then the element type would be int[][][]. But that is not actually the expected element type on the left hand side. Expected element type should be int. 

So in the above code it is derived by looking at the original type and the number of indexes in ArrayMapAccessExpr expression.